A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,506 @@
+other values. We considered the alternative of NACKing (i.e. rejecting)
+the CDS update when unsupported values are present, however it was rejected
+because of the difficulty in using such configuration in mixed deployments.
+
+The `override_host_status` value will be included in both the config of the
+`xds_override_host_experimental` policy and in the `DiscoveryMessage` inner
+message of the config of the `xds_cluster_resolver_experimental` policy.
+
+Also note that [common_lb_config][] field is incompatible with the new
+[load_balancing_policy][] field used for custom LB policies (see
+[A52: gRPC xDS Custom Load Balancer Configuration][grfc_a52]) which means
+custom LB policies and stateful session affinity features will be incompatible
+with each other. This will need to be fixed eventually to be able to use these
+two features together. Until that happens, we assume the default value
+for the [`common_lb_config.override_host_status`][or-host-status] field
+when one is using the new [load_balancing_policy][] field so the user is still
+able to use the feature.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel, the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing (including response processing) and just pass the
+RPC through (to the next filter). For example if the `path` configuration
+value is `/Package1.Service2` and the RPC method is
+`/Package2.Service1/Method3` then just pass the RPC through. Note, for path
+matching the rules described in [rfc6265 section-5.1.4][rfc6265-path-match]
+will apply.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
+set of all the cookies present. If the set does not contain a cookie whose
+name matches the `name` configuration value of the filter, then pass the RPC
+through. If you find more than one cookie with that `name` then log a local
+warning and pass the RPC through. Otherwise get the value of that cookie and
+base64-decode the value to get the `override-host` for the RPC. This value
+should be a syntactically valid IP:port address: if not, then log a local
```

Comment:
I do think we should try to have consistent behavior between Envoy and gRPC for these validations.

For the case of multiple cookies, what does Envoy do today?  Does it just use the first cookie it sees?  We could do the same thing in gRPC, or we could have Envoy ignore the cookie if it's a duplicate.  Which behavior seems better for users?

Logging a message when the format check fails is not strictly necessary in terms of correct behavior, but it may be useful for debugging purposes.  However, we'll have to make sure we don't flood the logs by logging a message for every individual RPC.

Reply:
>  Does it just use the first cookie it sees?

Yeap.

Reply:
Okay, then I think we should do the same thing here.  It's probably less-bad behavior anyway.

Reply:
Interesting, thanks for pointing that out.

I will note that the spec says "SHOULD NOT" instead of "MUST NOT", so technically we're not in violation if we do depend on the ordering.

Also, technically, the spec says that *servers* should not rely on the order, but in this case, it's the client that will be looking at it, not the server.  But I'm not sure whether that matters, because I don't know why the spec added that recommendation.  If the concern was that some intermediate proxy might have shuffled the original order of the headers, then depending on the order on the client side might be fine, but if it was for some other reason, I'm not sure what we'd break here.

In general, though, given this recommendation in the RFC, I think I agree that we should not depend on the ordering here.  So I think we have two options:

1. Pick one of the duplicates at random.
2. Ignore all of them and fall back to delegating to the child policy, as if there was no cookie set.

@ejona86 I'd like your input on this.

@wbpcode Whatever we decide here, we should change Envoy to do the same thing.

Reply:
Changed to say "pick the first cookie seen"

Reply:
The sentence _after_ the one quoted by @sanjaypujare is really the more useful sentence for this discussion, and answers @markdroth's wondering as to what the spec was concerned about:
> In particular,
   if the Cookie header contains two cookies with the same name (e.g.,
   that were set with different Path or Domain attributes), servers
   SHOULD NOT rely upon the order in which these cookies appear in the
   header.

So if we see duplicate headers, it is probably because the user changed the configuration, and we're seeing stale Cookies. But the cookie protocol itself won't be able to distinguish the stale from non-stale. I think we should probably make sure something about this gotcha is mentioned in user documentation.

If we see duplicates, we may be seeing them for a very long time. The time of duplication is completely dependent on the user-configured cookie age, which may be "as long as the client is running." For that reason, I don't think "choosing first" or "choosing randomly" provides much benefit as you will "randomly" not have session affinity and the non-determinism makes it harder to diagnose. Instead, I see the main choices as:
1. Ignore all the cookies, and accept that session affinity will be broken until the older cookie ages out (which might never happen)
2. Choose both cookies, passing down a list of allowed addresses to the LB. The LB would then choose the "first" address that has a subchannel available. Note that it would be exceedingly rare for more than one address to have a subchannel, because the second cookie wouldn't have been created if the existing cookie was working
3. Copy (at least) the domain and path into the cookie value, and ignore any cookies that don't match the current configuration. That _should_ leave us with one cookie

I would be very tempted to choose (1), but rolling back the configuration does not fix the problem. The user has to choose a new cookie name to resolve the problem, or wait for cookies to time out if an age was set.

Assuming option (2) is as easy to implement as it seems in my mind, that would be my preference. Option (3) is fine, but seems more prone to failure and bugs and cross-implementation drift.

Reply:
Eric, thanks for bringing up the point about a config change; I hadn't thought about that, but I agree that we need to be sure we're going to handle that correctly.  I think you're right that that's the one concrete case where duplicate cookies would actually happen in practice.

I agree that option 1 would cause breakage when the config changes, and rolling back the config would not fix it.  I think that's a good reason not to go with that option.

Option 3 seems like it would also cause problems for config changes.  If the config was changed to use a different path for the cookie (which I think is the only case that can trigger this problem) and that causes us to ignore existing cookies that have the wrong path, then the config change will basically break all in-progress sessions.

Option 2 seems like the best behavior, although it may be a little complicated to implement.  It will require passing a list instead of just a single string from the HTTP filter to the `override_host` LB policy, and it will slightly complicate the matching logic in the picker (basically, we'll need to loop over every cookie value looking for a subchannel, stopping at the first one that we find, or falling back to delegating to the child policy).  But I agree that it seems tractable.

@wbpcode, what do you think of changing Envoy to use option 2 here?

Reply:
@ejona86 and I chatted offline and he agreed that option 1 is acceptable with user documentation. If you look at a typical use-case of a user configuring session affinity for a service, this kind of config update etc is not going to be frequent and if we document that making such config change in the middle of ongoing sessions will impact session affinity that should be acceptable.

Reply:
Because I think option 2 would be rarely exercised so is not worth the additional complexity. In most cases users would just configure it once and leave it like that (unless they have made a mistake).

@va7ish please let me know what you think.

Reply:
My point is that there's almost no complexity involved in option 2, so even if it's hardly ever used, it seems like a better choice.

Reply:
In your [comment](https://github.com/grpc/proposal/pull/319#discussion_r1017236990) you mentioned "Option 2 seems like the best behavior, although it may be a little complicated to implement.... loop over every cookie value...". 

I agree with you it's all doable but requires some non-trivial changes in the design below. Let's hear from someone like @va7ish who intend to use this feature.

Reply:
Sorry, didn't mean to imply that I thought there was a lot of complexity here.  There is *some*, but not enough to really be that big of a problem.  I think it's worth doing.

Happy to get input from @va7ish though.

Reply:
If I understand the options correctly then I don't think either of them is great. From the user perspective SessionAffinity applies to LB decision and must be maintained (as long as the endpoint is valid) for the entire length of the session and independent of any new configuration changes. The state of the LB decision is stored inside the cookie and the lifetime of that decision is bound through cookie's TTL. Any configuration change should apply only to new LB decisions. If we encode enough contextual information inside the cookie we can later test if it is being misused (e.g. against a wrong path) by the client and ignore them. The logic should decode the cookie, verify that it is being used with a correct domain&path and pass the decoded endpoint down to the LB layer. A new cookie is going to be produced only when a new LB decision is made. The logic should probably ignore cookies with names that don't match the current configuration. If I understand things correctly name+path uniquely identifies a cookie. So whenever a new cookie is generated for a given name+path the client must replace the cookie in its "jar", correct? If we encode enough information in the cookie itself, the only breaking configuration change would be if the name of the cookie changes. And we can document that. 
Does this make sense? Am I missing something ?

Reply:
AIUI, cookie paths are basically [prefix matches, not full-match](https://www.rfc-editor.org/rfc/rfc6265#section-5.1.4).  So while it's true that a cookie is uniquely identified in the cookie jar using both the name and the path, there can be multiple paths that match a given request.

In practice, I think the case where this problem comes into play is when the config is changed such that the cookie uses the same name but a different path.  Let's say that you are using an RPC method called `/MyService/Foo` and your original cookie config created a cookie with the exact path `/MyService/Foo`.  But now you decide to add a second method, `/MyService/Bar`, and you want to use the same cookie for both methods -- i.e., within a single session, you will have some RPCs calling each of the two methods, but you want all of those RPCs to go to the same endpoint, regardless of which method they were sent to.  So you change the cookie config to use the same cookie name but to set the path to `/MyService` instead of `/MyService/Foo`, so that it will match both `/MyService/Foo` and `/MyService/Bar`.

Now let's say that you deploy that config change.  Any sessions that started prior to that receiving the updated config will be using a cookie with the original path, `/MyService/Foo`.  Because the path associated with the cookie is not sent on subsequent requests, just the cookie name and value, gRPC won't know on subsequent requests what the path of the cookie is, so it will keep using it.  However, if the original endpoint fails at this point, gRPC will pick a new endpoint and generate a new cookie, and the new cookie will have the path from the updated config, `/MyService`.  Since that path is not the same as the path that was generated for the original cookie, the cookie jar in the client application would not consider it to have the same ID as the original cookie, so it will store both of them, and both cookies will then show up on subsequent requests.  That's the point at which we need to figure out how to handle the duplicate cookies -- but note that at this point, the original endpoint had already failed, which is the reason we got the second cookie to begin with.

So the proposal here for option 2 is to go through the list of cookies and use the first one we find that actually works.  As long as the original endpoint is down, this should always pick the new endpoint and essentially ignore the original cookie.  However, if the original endpoint comes back up before the session ends, then it could cause some fluctuating behavior if the cookies are not sent in a stable order by the application.

In contrast, option 1 would completely break session affinity until either (a) the config is changed to use a different cookie name or (b) one of the cookies' TTLs expires and the application stops sending it.

Reply:
ok. I see. The effect of switching to a new endpoint, if the previously selected one is down, is very much application specific. In some cases it would just have a small initial performance impact, in others it would completely break the application. This really depends on the "hardness" of the state in the stateful service backend. If there is some fluctuation in the endpoint health state it would make matters even worse if requests get bounced back and forth across endpoints. 

We need to make sure that session-affinity processing logic is deterministic, assuming the last selected endpoint is available (its health state matches against the override_host_status). 

If we state that a "set-cookie" header is sent back(in the response metadata) if and only if a new LB decision was made then we can encode an internal revision id (incremented every time there is a config change or timestamp, etc) of the cookie (identified by its name) inside the value. The client cookie jar logic doesn't need to change in any way and would provide back all path-matching cookies it has on future requests. The affinity logic can then pick the value with the greatest revision-id as an override host for LB. This way we can at least guarantee some determinism such that once a switch over to a new endpoint has occurred there will be no flipping back to the previous endpoint unless another endpoint failure triggers new LB decision. 

Another option (not sure how clean that is) is to fail the request routing completely (with an error code) if the cookie provided leads to an endpoint that is completely down (doesn't match the override_host_status). This would allow the client to retry the request or empty the cookie-jar and submit the new request without any cookies.

Reply:
I like the revision (version) number idea: it only needs to be implemented in the `CookieBasedStatefulSessionFilter` and seems relatively easy to implement: for any RPC it needs to maintain the old revision number (in addition to the `upstreamAddress` value). In case a new cookie needs to be created because `hostAddress` (peer address) didn't match `upstreamAddress`, then it will bump up the revision number. Since only one cookie is selected on the request side, we don't need to pass multiple values in `override-host` to the LB policy.

The option in your last para is the easiest one to implement.

However in either case the behavior will be different from Envoy's current behavior. If we want to keep it same we'll need to co-ordinate.

Reply:
I'm open to the idea of encoding a version ID in the cookie value.  I'd want to define the range of the values and make sure we have the right behavior on wrap-around, just to make sure we don't hit any edge cases involving a lot of config churn, but that seems simple enough to do.

@ejona86, any concerns about this from your perspective?

@wbpcode, would you be open to changing Envoy to do the same thing?  We will need to define the exact format of the cookie value, so that gRPC and Envoy can interoperate.

I'd prefer not to just fail the RPC if the cookie exists but points to an unreachable endpoint.  I think that would require the application behavior and the xDS config to be more tightly coupled, which would make it harder to change the configuration over time.

Reply:
I think @sanjaypujare's counter approach would be good. But really I think we're well within the realm of "multiple working choices."

----

To make things more complicated for a sec, let's go back to Mark's example and cover a missed case: a cookie has `Path=/MyService/Foo` and is changed to `Path=/MyService` to share a session with `/MyService/Bar` as well. If only `/MyService/Foo` is accessed, then it progresses like I had said and Mark confirmed, where things only get weird if the backend goes down. But things get very unclear if `/MyService/Bar` is accessed, as it will set a conflicting cookie and both "work" and are "correct" for `/MyService/Foo`, depending on your point of view. If you are more interested in keeping the existing session you should use the `Path=/MyService/Foo` cookie. If you are more interested in a consistent session across Foo and Bar you should use the `Path=/MyService` cookie.

> Option 3 seems like it would also cause problems for config changes.

Yes and no. We could avoid the easy config changing case by only applying the cookie filtering if there is more than one cookie with the same name. If the config changes _again_, such there are 2 stale cookies, then it loses the ability to keep using the old session. But unclear how much we care about that (I don't think I care all that much).

> any concerns about this from your perspective?

For the revision ID approach to be workable for Envoy as well means we can't base the ID on any process-local state (like a local config counter). A client may be hitting multiple Envoy instances and today they all would route to the same backend.

----

Incrementing on each update (@sanjaypujare's suggestion) has some nice features, as it would work with multi-instance Envoy. Recognize it increases each time the address changes, not each time the config changes.

As far as ID wraparound, if we assume that the ID didn't wraparound at the same time as the config changed, then things are mostly fine. If there is only one cookie with the same configuration, then version 999 would just be replaced with version 0 and everything works perfectly. If there are some stale cookies, then version 999 would be replaced with version 0 but version 950 could exist. That would cause the next request to replace version 0 with version 951. (This case is possible to detect and avoid by noticing version 950 when resetting the id. Dunno how much we care.) If version 999 was a different configuration, then affinity will be broken until that cookie expires.

I think it is safe to say a max of 2^31 is plenty, as it won't be changing multiple times a second for many years. And if it _does_ change that frequently, we don't care about a single extraneous re-pick, because obviously affinity doesn't matter that much to the service. Even 2^24 seems fine.

Reply:
> To make things more complicated for a sec, let's go back to Mark's example and cover a missed case: a cookie has `Path=/MyService/Foo` and is changed to `Path=/MyService` to share a session with `/MyService/Bar` as well. If only `/MyService/Foo` is accessed, then it progresses like I had said and Mark confirmed, where things only get weird if the backend goes down. But things get very unclear if `/MyService/Bar` is accessed, as it will set a conflicting cookie and both "work" and are "correct" for `/MyService/Foo`, depending on your point of view. If you are more interested in keeping the existing session you should use the `Path=/MyService/Foo` cookie. If you are more interested in a consistent session across Foo and Bar you should use the `Path=/MyService` cookie.

Thanks for bringing this example up!  I hadn't considered that case, but I think it's significant, because it's a case where we would wind up with multiple cookies even when none of the endpoints are unreachable.  And then we would have two cookies with two different reachable endpoints to choose from.

@va7ish I'd like your thoughts on this use-case.  What behavior would you want to have happen here?

Note that a simple, monotonically increasing ID would not actually help in this case, because the new cookie created for the first request to `/MyService/Bar` would start with ID 0, which would be less than or equal to the ID on the pre-existing cookie for `/MyService/Foo`.  So even if we agreed that we wanted to prefer the cookie from the first request to `/MyService/Bar`, we wouldn't have a way to know that that one should be preferred.  This might push us in the direction of using some externally-generated state for the ID rather than a monotonically increasing counter.  (More on that below.)

One other thought: I think the existance of this use-case increases the importance of avoiding these edge cases whenever possible.  I might have been willing to live with config changes causing breakage only in the case that an endpoint went down right after the config change, because requiring both of those events to trigger the problem decreases the likelihood of anyone running into it.  But if it can happen without that second event, then I'm more worried about it.  In general, I think that however we handle these cases, the behavior is going to be sufficiently surprising to users that I'd like to prevent it if we can.

> > Option 3 seems like it would also cause problems for config changes.
> 
> Yes and no. We could avoid the easy config changing case by only applying the cookie filtering if there is more than one cookie with the same name. If the config changes _again_, such there are 2 stale cookies, then it loses the ability to keep using the old session. But unclear how much we care about that (I don't think I care all that much).

Interesting idea.  So the proposal here is to put the path in the cookie value but use that information only if there are more than one cookie and we need to choose between them.

I agree that this is better than the original idea for option 3.  I also agree that having more than one stale cookie is unlikely, but I'd still prefer to avoid this if possible.

> For the revision ID approach to be workable for Envoy as well means we can't base the ID on any process-local state (like a local config counter). A client may be hitting multiple Envoy instances and today they all would route to the same backend.

Good point.  We wouldn't have needed any such shared state if we could simply use a monotonically increasing counter, but as I mentioned above, that won't work in the use-case you just brought up.  So we probably do need to solve this somehow.

One option would be to use a timestamp.  However, that would not necessarily do the right thing if there is clock skew between different data planes, so I'm not sure it's something we should rely on.

Another option is that we could have the xDS config include the current ID and an ordered list of old IDs, so that we always know which cookies are newer whenever there are more than one.  That way, we know it would be consistent no matter what the data plane is.  This would put control into the user's hands, but we wouldn't really be able to provide any sanity check if they get the config wrong.

@va7ish, would it be useful to you to have that kind of mechanism in the configuration?

> Incrementing on each update (@sanjaypujare's suggestion) has some nice features, as it would work with multi-instance Envoy. Recognize it increases each time the address changes, not each time the config changes.

Yeah, I think we'd definitely want to do this every time the address changes, not every time the config changes.  Doing it every time the config changes wouldn't really help, because the data plane doesn't know the ordered list of config versions.

Reply:
> To make things more complicated for a sec, let's go back to Mark's example and cover a missed case: a cookie has `Path=/MyService/Foo` and is changed to `Path=/MyService` to share a session with `/MyService/Bar` as well. If only `/MyService/Foo` is accessed, then it progresses like I had said and Mark confirmed, where things only get weird if the backend goes down. But things get very unclear if `/MyService/Bar` is accessed, as it will set a conflicting cookie and both "work" and are "correct" for `/MyService/Foo`, depending on your point of view. If you are more interested in keeping the existing session you should use the `Path=/MyService/Foo` cookie. If you are more interested in a consistent session across Foo and Bar you should use the `Path=/MyService` cookie.

The sharing of sessions between `/MyService/Foo` and `/MyService/Bar` should only occur on new sessions created after the config changed. 
If the configuration changes from `/MyService/Foo` to `/MyService` then a new cookie value gets generated only when either of the two happens : 1) existing endpoint is not usable 2) The client is accessing another method that doesn't have a cookie value generated for yet.

When #1 happens then the session affinity is essentially broken (forced as such) and a new session mapping is created. This leads to a new cookie value but the underlying cookie name is the same so the internal version id can be incremented and encoded in the value. 

When #2 happens, essentially this is a new session from the perspective of the logic we are discussing here. Client will be given a new cookie value (with a broader path match) for that new session and it will be added to the cookie jar for future calls. 
What we have is that client is going to be sending two session cookies (the new and the old) when it calls `/MyService/Foo` (because of path matching) but only a single cookie (the new cookie) when it calls `/MyService/Bar`. So we need to define the algorithm for properly maintaining the original session affinity when `/MyService/Foo` is called as in the example here. If we encode the `Path` portion of the config in the cookie value when we generate the cookie then the logic could be that the cookie with the longest path match should be used first as long as the previously selected endpoint is still valid. 

If the configuration is changed in the opposite direction (i.e. from broad to specific path) then we don't have a problem as the new cookie will only be generated when the previous session mapping failed or expired.


> Note that a simple, monotonically increasing ID would not actually help in this case, because the new cookie created for the first request to /MyService/Bar would start with ID 0, which would be less than or equal to the ID on the pre-existing cookie for /MyService/Foo. So even if we agreed that we wanted to prefer the cookie from the first request to /MyService/Bar, we wouldn't have a way to know that that one should be preferred. This might push us in the direction of using some externally-generated state for the ID rather than a monotonically increasing counter. 
I am not sure I follow why the new cookie config (with same `name` but different `path`) would lead to ID resetting to 0.

Reply:
I think this discussion has gone on for too long, and also we have not heard from @wbpcode for the Envoy angle, so let me just propose my changes (I also had an offline discussion with @ejona86):

In the cookie we'll encode the path (that was configured for the cookie) when the cookie is injected into an RPC response. We will *not* encode any revision id or number. While processing an RPC on the request side, we only pick matching cookies if there are multiple cookies: exact match for name and prefix match for the path (both ways - the configured path can be a prefix of the cookie path or the cookie path can be a prefix of the configured path).

Then we do further filtering as follows: 

- the path in the cookie has to match the current RPC (obviously) e.g. if the current RPC is /MyService/Foo and the cookie has /MyService/Bar then we discard the cookie. Ideally the client should not send such a cookie but with the path encoded in the cookie we can at least do the check.
- if we are still left with multiple cookies and one of the cookies has an exact match with the currently configured cookie path then pick that cookie e.g. with 2 cookies /MyService/Foo and /MyService if the current config is /MyService/Foo then pick the cookie that has /MyService/Foo but if the current configuration is /MyService then pick the cookie that has /MyService (@va7ish this is different from your suggestion of picking the cookie with the longest path - the logic is we are likely to pick the most recent cookie with my proposed logic)
- if there is no exact match (above bullet) then pick the cookie that has the longest path. This may not be the latest cookie but this is no worse than, say picking the first cookie.

Let me know if you have any objections. I would like to wrap this up as this seems to be the only outstanding issue.

Reply:
The main objective here is that a configuration change should not break any existing working session affinity. Such change should only affect new sessions. 

> * if we are still left with multiple cookies and one of the cookies has an exact match with the currently configured cookie path then pick that cookie e.g. with 2 cookies /MyService/Foo and /MyService if the current config is /MyService/Foo then pick the cookie that has /MyService/Foo but if the current configuration is /MyService then pick the cookie that has /MyService (@va7ish this is different from your suggestion of picking the cookie with the longest path - the logic is we are likely to pick the most recent cookie with my proposed logic)

What you are suggesting will break existing sessions in the following scenario:

1. Cookie is configured for `/MyService/Foo`
2. `/MyService/Foo` is called without a cookie, designating a new session. New cookie is created (`/MyService/Foo` is encoded inside) and returned to the client
3. Configuration is changed to `/MyService`. 
4. `/MyService/Bar` is called without a cookie, designating a new session. New cookie is created (`/MyService` is encoded inside) and returned to the client. 
5. At this point client's cookie jar has two cookies: `/MyService/Foo` and `/MyService`. 
6. Client calls `/MyService/Foo` and supplies both cookies from its jar. 
7. The logic you propose will select the second session path instead of the original one that was selected for `/MyService/Foo` in step 2. This breaks existing healthy session affinity. 

Please let me know if I have missed something in your proposal.

Reply:
The example I was thinking of is the following sequence of events:

1. Initial cookie configuration is /MyService/Foo
2. /MyService/Foo is called and a cookie C1 is created with /MyService/Foo encoded in the cookie
3. Now the cookie config is changed to /MyService 
4. The backend encoded in C1 is unreachable so gRPC creates a new cookie C2 with /MyService in the cookie
5. Next call to /MyService/Foo will have both C1 and C2. If gRPC picks C1 because it has the longest path it will use that backend if it's now reachable. But that would be the wrong choice given the semantics, isn't it?

But I see the point with your example which applies even when no backends go down. One way is to also include the the revision number and use both the number and the cookie path. The logic in that case will be as follows:

1. When a cookie is created include both the current configured cookie path and a "revision number". The revision number is bumped up (from the current cookie) only if the RPC was sent to a backend different from what the current cookie indicated.
2. When we receive multiple cookies on the request side, we pick the cookie C0 with the longest matching path (and with the highest revision number if there are multiple such cookies). However if we find a cookie C1 with a higher revision number and a matching path (but shorter than the one in C0) then we prefer that cookie.

In your example (with no unreachable backends) both cookies will have the same revision number ( say 1) and picking the cookie with the longest matching path will be correct. In my example (which has both a config change and a failing backend after config change) picking a cookie with a higher revision number (but with a shorter prefix matching path) will still be correct.

What do you think?

Reply:
> 5\. Next call to /MyService/Foo will have both C1 and C2. If gRPC picks C1 because it has the longest path it will use that backend if it's now reachable. But that would be the wrong choice given the semantics, isn't it?

I don't think that's the wrong semantics. C1 represents a session (with a given TTL) that needs to be maintained if the endpoint is up. 


> When we receive multiple cookies on the request side, we pick the cookie C0 with the longest matching path (and with the highest revision number if there are multiple such cookies).

AFAIK From https://www.rfc-editor.org/rfc/rfc6265#section-5.3 a cookie is uniquely identified by [name, domain and path]. Client can't send more than one cookie with the same name and path (encoded inside the value).

Reply:
> > 5. Next call to /MyService/Foo will have both C1 and C2. If gRPC picks C1 because it has the longest path it will use that backend if it's now reachable. But that would be the wrong choice given the semantics, isn't it?
> 
> I don't think that's the wrong semantics. C1 represents a session (with a given TTL) that needs to be maintained if the endpoint is up.

Let me expand the sequence a bit:

1. Initial cookie configuration is /MyService/Foo
2. /MyService/Foo is called and a cookie C1 is created with /MyService/Foo encoded in the cookie
3. Now the cookie config is changed to /MyService
4. The backend encoded in C1 - let's call it B1 - becomes unreachable
5. The next RPC from the client has the cookie C1 but B1 is unreachable, so gRPC falls back to regular load balancing and routes the RPC to a new backend B2 and allocates a new cookie C2 that corresponds to B2. Note, C2 has the path /MyService now. Also note that as far as the client is concerned the session is now established with B2.
6. Client includes both cookies C1 and C2 in subsequent RPCs and as long as B1 is unreachable and B2 is reachable the RPCs will be routed to B2 (which is correct?).
7. Now C1 becomes reachable.
8. The client sends the next RPC in the current session (which has both C1 and C2) which will now be routed to B1 if we pick the cookie that has the longest path and this in my opinion is not what you expect.
 
Let me know what you think.

> 
> > When we receive multiple cookies on the request side, we pick the cookie C0 with the longest matching path (and with the highest revision number if there are multiple such cookies).
> 
> AFAIK From https://www.rfc-editor.org/rfc/rfc6265#section-5.3 a cookie is uniquely identified by [name, domain and path]. Client can't send more than one cookie with the same name and path (encoded inside the value).

You are right. However note that the path is in 2 places: the [path attribute](https://www.rfc-editor.org/rfc/rfc6265#section-5.2.4) in the Set-Cookie header and also encoded inside the cookie value (as per our proposed logic) and ideally they should be the same. If they are not the same for whatever reason (the client has a bug - let's say they use their own cookie jar implementation which stores multiple cookies with the same path), I just wanted to be prepared for that possibility.

Reply:
Yes, I agree that flip-flopping the session routing between (6) and (8) is bad. 

> (1) When a cookie is created include both the current configured cookie path and a "revision number". The revision number is bumped up (from the current cookie) only if the RPC was sent to a backend different from what the current cookie indicated.
(2) When we receive multiple cookies on the request side, we pick the cookie C0 with the longest matching path (and with the highest revision number if there are multiple such cookies). However if we find a cookie C1 with a higher revision number and a matching path (but shorter than the one in C0) then we prefer that cookie.

My concern here is about revision number scoping. Consider the following sequence:

1. The cookie C1 that was generated for `/MyService/Foo` routes to B1 properly. Rev = 1
2. Config change occurs and the cookie path matching for new sessions should now be `/MyService`. Client calls `/MyService/Bar` and supplies no cookies (because the path it has for C1 doesn't match `/MyService/Bar`. New cookie cookie C2, Rev=1 gets generated. 
3. Client calls `/MyService/Bar`with C2 again but this time B2 is down. So a new B3 is selected, C2 is updated with new revision Rev=2, correct? 
4. Now Client calls `/MyService/Foo` and supplies C1 and C2. B1 is still healthy and available to handle traffic. If I understand your endpoint selection algorithm correctly then C2 will be used (as it has higher revision) leading to B3 being used for this RPC instead of B1, which will break C1->B1 session affinity. Correct?

perhaps when step (6) occurs, in your sequence you mentioned above, we need to respond back with an updated C1 value as well which would indicate that `/MyService/Foo` calls are also explicitly routed to B2 from now on. 

One aspect that makes me wary about this is the concurrency element. Client might be asyncrhonously sending/queuing multiple RPC requests to `/MyService/Foo` and some will be queued with C1 only while others will have C1 and C2.

Reply:
> ...
> perhaps when step (6) occurs, in your sequence you mentioned above, we need to respond back with an updated C1 value as well which would indicate that `/MyService/Foo` calls are also explicitly routed to B2 from now on.

On the response path, the code will only look at the cookie for the current configuration. In this example the current configuration is `/MyService` whereas C1 was for the path `/MyService/Foo` so why would the code update C1? What would be the logic to allow the code to update unrelated cookies? In this case `/MyService` happens to be a prefix of `/MyService/Foo` (so more generalized config) but we will need to codify the logic. 

I am thinking we will always find examples that will break no matter how much we tweak the logic. So I suggest we settle on the logic described [above](https://github.com/grpc/proposal/pull/319#discussion_r1026749828).

> 
> One aspect that makes me wary about this is the concurrency element. Client might be asyncrhonously sending/queuing multiple RPC requests to `/MyService/Foo` and some will be queued with C1 only while others will have C1 and C2.

I think that's inherent with the cookie approach here so I am not sure anything can be done.

Reply:
> In this case /MyService happens to be a prefix of /MyService/Foo (so more generalized config) but we will need to codify the logic.

Yes, I think we would need to do that. 

> I am thinking we will always find examples that will break no matter how much we tweak the logic. So I suggest we settle on the logic described https://github.com/grpc/proposal/pull/319#discussion_r1026749828.

Well, the main point is that we can't break existing session affinity if the backend is healthy even when the configuration is changed. The only configuration change that needs to be considered (as it is the only one that would break SSA) is from a specific path to more generalized path (`/MyService/Foo` -> `/MyService`)

Reply:
> Well, the main point is that we can't break existing session affinity if the backend is healthy even when the configuration is changed.

Is that really true? As an example if the config is completely changed e.g. from `/MyService/Foo` to `/AnotherService/Bar` then the `CookieBasedStatefulSessionFilter` filter will (only) honor the new config and hence will ignore cookies that don't match the current config i.e. session affinity will be broken with old cookies.

 > The only configuration change that needs to be considered (as it is the only one that would break SSA) is from a specific path to more generalized path (`/MyService/Foo` -> `/MyService`)

My main point is that when the filter is operating with the new configuration (`/MyService`), it should not attempt to update cookies that have a different path (even if a subpath) for a variety of reasons (may be the cookie was issued by a different entity, we don't really want to guess what the user really wants). We can use a cookie with a "compatible" path but should not update it. 

> > In this case /MyService happens to be a prefix of /MyService/Foo (so more generalized config) but we will need to codify the logic.
> 
> Yes, I think we would need to do that.

Let me think some more. Or what I would like to suggest is to address some of these corner or remote cases as an update of the current spec after we have had a chance to test the current implementation.

Reply:
> Is that really true? As an example if the config is completely changed e.g. from `/MyService/Foo` to `/AnotherService/Bar` then the `CookieBasedStatefulSessionFilter` filter will (only) honor the new config and hence will ignore cookies that don't match the current config i.e. session affinity will be broken with old cookies.

Yes. In this example above, the parent services are different so I don't think the user will have strong expectations with regards to retaining session affinity. But I agree that it would be nice to retain SSA even in such cases. 


> Let me think some more. Or what I would like to suggest is to address some of these corner or remote cases as an update of the current spec after we have had a chance to test the current implementation.

I am good with resolving this in phases. But at least we need to be very explicit with regards to which configuration changes are supported (will not not break SSA).

Reply:
So the decision (for now) is to honor the first cookie we see. The reason for this is that the client then sees the same behavior regardless of whether the mode is proxyless or proxy-based and is easier to reason about.

At a later stage we will update both the Envoy and gRPC implementations to have the same behavior which will be collaboratively decided.